### ISSUE TITLE
Consider checking for -eq used with strings

### ISSUE THREAD

**Issue by koalaman** [koalaman/shellcheck#147]

**Comment by deryni** [comment]
This doesn't seem to be working (assuming the intention is to warn when a variable with a string value is compared using `-eq`)

The following script passes without comment on shellcheck.net.

    state=IN_INIT
    [[ $state -eq 0 ]]
    echo $?
    [ $state -eq 0 ]
    echo $?

This is particularly unfortunate with `[[` since that, apparently, expands the variable and then evaluates the resulting word in arithmetic context and ends up using `0` as the value (for undefined variables).

**Comment by koalaman** [comment]
This is true -- the fix only triggers when comparing against literal non-integers. 

It's tricky since it's hard to determine whether an integer contains a number in a way that isn't intrusive, especially considering things like `state=$(grep state file | cut -d ':' -f 2)`. 

Reopening to look into whether easily detectable literal string examples like this one are frequent, and whether there is more that could be done to differentiate int from non-int variables in a way that doesn't flood with false positives.

**Comment by cxw42** [comment]
Another test case:

    #!/bin/bash
    testme=value
    criterion=value
    [[ testme -eq 'criterion' ]] && echo yep

reports no errors.  However, when run in bash 4.4.12(3), it outputs `yep`, which is not likely the author's intent :) .  I appreciate your point about the difficulty of the general case.  Perhaps flag `/-eq\s\+'/` and `/-eq\s\+"[^$]/` as a start?

**Comment by kurahaupo** [comment]
My suggestion would be to recommend reserving [[ ]] for strings and related
operations, and instead recommend (( )) for all numeric operations including
comparisons.

(That is, treating comparators as just more arithmetic operators, as the C
language does, but which many Shell guides do not.)

Some shell guides already recommend using (( )) for arithmetic, but fall
short of explaining that, as far as C (and C++ and Perl and Python and PHP
and JS etc) are concerned, comparisons ARE arithmetic, along with bitwise
ops and bitshifts; that leaves the reader presuming that [[ ]] should be
used for comparisons, even when numbers are intended. So -eq and -ne are
strangely popular.

Which would mean effectively treating the -eq, -ne, -gt, -ge, -lt & -le
operators as if they were deprecated, and issuing warnings that recommend
replacing them with the corresponding C-like operators within (( )).

If that seems a step too far, a less intrusive possibility would be to raise
a warning along the lines of "quoted string is not literal in numeric
context". (There's no globbing or word-splitting there, so it doesn't matter
whether $ or * are present.)
